File: library/src/main/java/org/incendo/interfaces/interfaces/AbstractInterfaceBuilder.java
package org.incendo.interfaces.interfaces;

import org.bukkit.event.inventory.InventoryCloseEvent;
import org.incendo.interfaces.click.ClickHandler;
import org.incendo.interfaces.pane.Pane;
import org.incendo.interfaces.properties.Trigger;
import org.incendo.interfaces.transform.AppliedTransform;
import org.incendo.interfaces.transform.ReactiveTransform;
import org.incendo.interfaces.transform.Transform;
import org.incendo.interfaces.utilities.IncrementingInteger;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.ArrayList;
import java.util.HashSet;

public abstract class AbstractInterfaceBuilder<P extends Pane, I extends Interface<P>> implements InterfaceBuilder<P, I> {

    private static final List<InventoryCloseEvent.Reason> DEFAULT_REASONS = new ArrayList<>();

    static {
        for (InventoryCloseEvent.Reason reason : InventoryCloseEvent.Reason.values()) {
            if (reason != InventoryCloseEvent.Reason.PLUGIN) {
                DEFAULT_REASONS.add(reason);
            }
        }
    }

    private final IncrementingInteger transformCounter = new IncrementingInteger();

    private Map<InventoryCloseEvent.Reason, CloseHandler> closeHandlers = new HashMap<>();
    private Collection<AppliedTransform<P>> transforms = new ArrayList<>();
    private Collection<ClickHandler> clickPreprocessors = new ArrayList<>();

    public final Map<InventoryCloseEvent.Reason, CloseHandler> closeHandlers() {
        return this.closeHandlers;
    }

    public final void closeHandlers(final Map<InventoryCloseEvent.Reason, CloseHandler> closeHandlers) {
        this.closeHandlers = closeHandlers;
    }

    public final Collection<AppliedTransform<P>> transforms() {
        return this.transforms;
    }

    public final void transforms(final Collection<AppliedTransform<P>> transforms) {
        this.transforms = transforms;
    }

    public final Collection<ClickHandler> clickPreprocessors() {
        return this.clickPreprocessors;
    }

    public final void clickPreprocessors(final Collection<ClickHandler> clickPreprocessors) {
        this.clickPreprocessors = clickPreprocessors;
    }

    public final void withTransform(final Transform<P> transform, final Trigger... triggers) {
        this.transforms.add(new AppliedTransform<>(
            this.transformCounter.next(),
            new HashSet<>(List.of(triggers)),
            transform
        ));
    }

    public final void addTransform(final ReactiveTransform<P> reactiveTransform) {
        this.transforms.add(new AppliedTransform<>(
            this.transformCounter.next(),
            Set.of(reactiveTransform.triggers()),
            reactiveTransform
        ));
    }

    public final void withCloseHandler(final Collection<InventoryCloseEvent.Reason> reasons, final CloseHandler closeHandler) {
        for (InventoryCloseEvent.Reason reason : reasons) {
            this.closeHandlers.put(reason, closeHandler);
        }
    }

    public final void withCloseHandler(final CloseHandler closeHandler) {
        this.withCloseHandler(DEFAULT_REASONS, closeHandler);
    }

    public final void withPreprocessor(final ClickHandler handler) {
        this.clickPreprocessors.add(handler);
    }

}

File: next/src/main/java/org/incendo/interfaces/next/interfaces/AbstractInterfaceBuilder.kt
package org.incendo.interfaces.next.interfaces

import org.bukkit.event.inventory.InventoryCloseEvent
import org.incendo.interfaces.next.click.ClickHandler
import org.incendo.interfaces.next.pane.Pane
import org.incendo.interfaces.next.properties.Trigger
import org.incendo.interfaces.next.transform.AppliedTransform
import org.incendo.interfaces.next.transform.ReactiveTransform
import org.incendo.interfaces.next.transform.Transform
import org.incendo.interfaces.next.utilities.IncrementingInteger

public abstract class AbstractInterfaceBuilder<P : Pane, I : Interface<P>> internal constructor() :
    InterfaceBuilder<P, I>() {

    private companion object {
        private val DEFAULT_REASONS = InventoryCloseEvent.Reason.values().toList().minus(InventoryCloseEvent.Reason.PLUGIN)
    }

    private val transformCounter by IncrementingInteger()

    protected val closeHandlers: MutableMap<InventoryCloseEvent.Reason, CloseHandler> = mutableMapOf()
    protected val transforms: MutableCollection<AppliedTransform<P>> = mutableListOf()
    protected val clickPreprocessors: MutableCollection<ClickHandler> = mutableListOf()

    public fun withTransform(vararg triggers: Trigger, transform: Transform<P>) {
        transforms.add(AppliedTransform(transformCounter, triggers.toSet(), transform))
    }

    public fun addTransform(reactiveTransform: ReactiveTransform<P>) {
        transforms.add(AppliedTransform(transformCounter, reactiveTransform.triggers.toSet(), reactiveTransform))
    }

    public fun withCloseHandler(
        reasons: Collection<InventoryCloseEvent.Reason> = DEFAULT_REASONS,
        closeHandler: CloseHandler
    ) {
        reasons.forEach {
            closeHandlers[it] = closeHandler
        }
    }

    public fun withPreprocessor(handler: ClickHandler) {
        clickPreprocessors += handler
    }
}
